src/main/java/com/hivemq/persistence/InMemoryProducerQueuesImpl.java

```diff
@@ -0,0 +1,243 @@
+                                          @NotNull final Task<R> task) {
+        //noinspection ConstantConditions (futuer is never null if the callbacks are null)
+        return submit(bucketIndex, task, null, null);
+    }
+
+
+    public <R> @Nullable ListenableFuture<R> submit(final int bucketIndex,
+                                          @NotNull final Task<R> task,
+                                          @Nullable final SingleWriterService.SuccessCallback<R> successCallback,
+                                          @Nullable final SingleWriterService.FailedCallback failedCallback) {
+
+        return submit(bucketIndex, task, successCallback, failedCallback, false);
+    }
+
+
+    private <R> @Nullable ListenableFuture<R> submit(final int bucketIndex,
+                                           @NotNull final Task<R> task,
+                                           @Nullable final SingleWriterService.SuccessCallback<R> successCallback,
+                                           @Nullable final SingleWriterService.FailedCallback failedCallback,
+                                           final boolean ignoreShutdown) {
+        if (!ignoreShutdown && shutdown.get() && System.currentTimeMillis() - shutdownStartTime > shutdownGracePeriod) {
+            return SettableFuture.create(); // Future will never return since we are shutting down.
+        }
+        final int queueIndex = bucketIndex / bucketsPerQueue;
+        final SettableFuture<R> resultFuture;
+        if (successCallback == null) {
+            resultFuture = SettableFuture.create();
+        } else {
+            resultFuture = null;
+        }
+
+        final MpscUnboundedArrayQueue<Runnable> queue = inMemorySingleWriter.queues[queueIndex];
+        final AtomicInteger wip = inMemorySingleWriter.wips[queueIndex];
+        queue.offer(() -> {
+            try {
+                final R result = task.doTask(bucketIndex, queueBucketIndexes.get(queueIndex), queueIndex);
+                if (resultFuture != null) {
+                    resultFuture.set(result);
+                } else {
+                    successCallback.afterTask(result);
+                }
+            } catch (final Exception e) {
+                if (resultFuture != null) {
+                    resultFuture.setException(e);
+                } else {
+                    if (failedCallback != null) {
+                        failedCallback.afterTask(e);
+                    }
+                }
+            }
+        });
+
+        if (wip.getAndIncrement() == 0) {
+            int missed = 1;
+            do {
+                while (true) {
+                    final Runnable runnable = queue.poll();
+                    if (runnable != null) {
+                        runnable.run();
+                    } else {
+                        break;
+                    }
+                }
+                missed = wip.addAndGet(-missed);
+            } while (missed != 0);
+        }
```

Comment:
Maybe a small comment, that this *is* the described trampoline approach of `InMemorySingleWriterImpl` or that this is the main difference to `ProducerQueues`?